File: homework/thinhbka/lesson-6-binary-search-bst/450. Delete Node in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        parent = {}
        def find(node,key):
            if node.val == key:
                return node
            else:
                if node.val < key:
                    if node.right:
                        parent[node.right] = node
                        return find(node.right)
                    return None
                else:
                    if node.left:
                        parent[node.left] = node
                        return find(node.left)
                    return None
        def dfs(node):
            if node.left:
                parent[node.left] = node
                dfs(node.left)
            if node.right:
                parent[node.right] = node
                dfs(node.right)

File: homework/Truong_Duong/Lesson_9/450_Delete_Node_in_a_BST.js
// Solution for https://leetcode.com/problems/delete-node-in-a-bst/

var findRightMostNode = (node) => {
  if (node.right === null) return node;
  return findRightMostNode(node.right);
}

var deleteNode = function (root, key) {
  if (root == null) return null;
  
  if (key < root.val) {
    root.left = deleteNode(root.left, key);
  } else if (key > root.val) {
    root.right = deleteNode(root.right, key);
  } else {
    if (root.left == null) return root.right;
    if (root.right == null) return root.left;

    const rightMostNode = findRightMostNode(root.left);
      
    root.val = rightMostNode.val;
    root.left = deleteNode(root.left, root.val);
  }

  return root;
};